### ISSUE TITLE
[Bug][simpler][a5sim] Pass the simulated target to PTO ISA when compiling CPU kernels

### ISSUE THREAD

**Issue by Crystal-wzy** [hw-native-sys/simpler#2215]
### Problem

Selecting `platform="a5sim"` leaves the kernel worker's PTO ISA CPU memory
model at its default, `NPUArch::A2A3`. As a result, `TROWSUM` executes the
A2A3 reduction path even though the selected platform is A5.

Expected: simpler passes the simulated target when compiling the kernel;
PTO ISA selects that target's instruction semantics and memory configuration.
Observed: a fixed FP32 row-sum input returns **1 instead of 0**. Initializing
the worker as A5 fixes the result without changing the reduction or ISA headers.
In attention models, this difference can propagate through BF16 rounding
to cache and compressor-state outputs.

The proposed production fix uses a compile-time target parameter. The
runtime-initialization control below establishes the cause of the failure;
it is not the proposed integration mechanism.

### Environment

| Component | Revision / version |
| --- | --- |
| pypto-lib | `d6f2920046df7a19c28a5079c096a53b7e600219` |
| PyPTO | `8ecc5d19f0ac1062ae6779ba8aca8cee75c22667` |
| simpler | `39ce891dbb3f665e72e99b4a1387b47012fb18bd` + PR #2186 patches listed below |
| pto-isa | `bc1bd8b9544502a352c6cec1776b9ef00a8303f4` |
| PTOAS | `v0.61` |
| PyTorch | `2.6.0+cpu` |
| Python | `3.10.19` |
| CPU compiler | GCC 15.2.1, C++23, `-O2`, without `-ffast-math` |

The two simpler PR #2186 patches are
`14a46178d83ad31c69ae06adacad2113325206a0` and
`a9a149a4033cd49788582e9f109167c148b00579`.
These are CPU simulator tests; no NPU or model weights are required.

### Error case

From the pypto-lib repository root, with the dependencies above installed:

```bash
export PYTHONPATH=".${PYTHONPATH:+:$PYTHONPATH}"
export DEEPSEEK_V4_VARIANT=pro
timeout --kill-after=10s 300s python \
  models/deepseek_v4_pro/decode_attention_csa.py -p a5sim
timeout --kill-after=10s 300s python \
  models/deepseek_v4_pro/decode_attention_swa.py -p a5sim
```

| Model | Outputs affected by this issue |
| --- | --- |
| `decode_attention_csa.py` | `compress_state`, `inner_compress_state`, `cmp_kv` |
| `decode_attention_swa.py` | `kv_cache` |

These entrypoints generate random data, so they may pass on some runs.
Model controls below replayed captured failing inputs with the same golden
data and tolerances. The minimal reproduction uses fixed data and needs
no captured files. SWA's separate `x_out` failure is covered by
`subissue_pypto_missing_gm_cross_core_sync.md`.

### Minimal reproduction

Save this as `repro_simpler_a5_arch.py` in the pypto-lib repository root.
PyPTO generates a single row-sum kernel and launches it through simpler;
`golden` is pypto-lib's validation helper.

Each of the eight input rows contains `[1e8, 1, -1e8, 1]` repeated 16 times.
For this input, A5's modeled FP32 adjacent-pair reduction yields zero; the
observed A2A3 path yields one with the listed compiler flags. Eight rows
satisfy the generated reduction tile's 32-byte column alignment requirement.

The optional `--initialize-a5` control inserts A5 initialization at the
generated kernel entry, before Tile binding. It is a diagnostic hook for
the listed simpler version; it leaves the input, reduction, golden and
tolerances unchanged.

```python
"""Reproduce a5sim failing to select the PTO ISA A5 reduction path."""
import argparse
from pathlib import Path
import re

import torch
import pypto.language as pl
from golden import TensorSpec, run


@pl.jit
def row_sum(x: pl.Tensor[[8, 64], pl.FP32],
            out: pl.Out[pl.Tensor[[8, 1], pl.FP32]]):
    with pl.spmd(1):
        idx = pl.tile.get_block_idx()
        out[:, :] = pl.row_sum(x[:, :])
    return out


def golden(tensors):
    # Expected from the A5 FP32 adjacent-pair reduction for this exact input.
    tensors["out"].zero_()


def initialize_a5_for_diagnosis():
    from simpler_setup.kernel_compiler import KernelCompiler

    original_compile = KernelCompiler._compile_incore_sim

    def compile_with_a5(self, source_path, *args, **kwargs):
        path = Path(source_path)
        source, count = re.subn(
            r"(void kernel_entry\([^\n]*\)\s*\{)",
            r"\1\n    pto::NPUMemoryModel::Instance().Initialize(pto::NPUArch::A5);",
            path.read_text(),
        )
        assert count == 1, "Expected one generated kernel entry"
        diagnostic = path.with_name(path.stem + ".a5-arch.cpp")
        diagnostic.write_text(source)
        return original_compile(self, str(diagnostic), *args, **kwargs)

    KernelCompiler._compile_incore_sim = compile_with_a5


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--initialize-a5", action="store_true")
    args = parser.parse_args()
    if args.initialize_a5:
        initialize_a5_for_diagnosis()
    result = run(
        fn=row_sum,
        specs=[
            TensorSpec("x", [8, 64], torch.float32,
                       init_value=lambda: torch.tensor([1e8, 1.0, -1e8, 1.0]).repeat(8, 16)),
            TensorSpec("out", [8, 1], torch.float32),
        ],
        golden_fn=golden,
        config={"platform": "a5sim"},
        atol=0,
        rtol=0,
    )
    if result.error:
        print(result.error)
    print(f"passed={result.passed}")
    raise SystemExit(0 if result.passed else 1)
```

Run the two modes in separate processes. If an external compilation cache
is enabled, use fresh outputs so both modes compile their kernel entry.

```bash
export PYTHONPATH=".${PYTHONPATH:+:$PYTHONPATH}"
python repro_simpler_a5_arch.py
echo "rc=$?"
python repro_simpler_a5_arch.py --initialize-a5
echo "rc=$?"
```

Both modes use `atol=rtol=0`. Recorded results:

| Mode | Expected output | Actual output | Validation | Exit code |
| --- | --- | --- | --- | --- |
| Default `a5sim` | Eight zeros | Eight ones | 8/8 mismatches | 1 |
| `a5sim --initialize-a5` | Eight zeros | Eight zeros | PASS | 0 |

### Root cause and ownership

The failure follows this initialization path:

1. In simpler, `Gxx15Toolchain` (`simpler_setup/toolchain.py`) sets CPU_SIM
   and AIC/AIV compilation flags. `KernelCompiler._compile_incore_sim`
   (`simpler_setup/kernel_compiler.py`) and the tested simulator runtime
   do not select the A5 memory model for the executing worker.
2. In pto-isa, `include/pto/cpu/NPUMemoryModel.hpp` provides thread-local
   instances and defaults to `NPUArch::A2A3` when none is explicitly selected.
3. `include/pto/cpu/TRowSum.hpp` chooses its reduction path using that
   instance's `GetArch()`. Consequently, this A5 launch uses the A2A3 path.

The A5 reduction algorithm was already fixed in pto-isa
`9aba183fbb76f758272453ea5341e0c6185fa169`.
The remaining integration must enable it. The existing ISA contract in
`docs/coding/cpu_sim.md`, under "Selecting the simulated architecture",
requires external integrations to select the target before running kernels.
Together with the initialization-only control, this identifies the missing
target selection. The compile-time solution requires coordinated changes:
simpler must pass the target, and pto-isa must recognize it while retaining
the CPU backend. This issue tracks the simpler integration and specifies
the companion ISA interface it depends on.

### Proposed fix

Select the simulated architecture family at compile time using
`PTO_CPU_SIM_TARGET_A2A3` or `PTO_CPU_SIM_TARGET_A5`. simpler already knows
this family from `a2a3sim` / `a5sim`; it does not know which specific A5
variant is being simulated. No new user configuration is required.

`__CPU_SIM` continues to select the CPU execution backend, while
`__DAV_CUBE__` / `__DAV_VEC__` select the core type. The following interface
is proposed, not yet implemented.

#### 1. Compile-time interface

| simpler platform | Required family flag | ISA behavior |
| --- | --- | --- |
| `a2a3sim` | `-DPTO_CPU_SIM_TARGET_A2A3` | Compile the A2A3 CPU instruction paths and use the existing A2A3 simulator memory profile |
| `a5sim` | `-DPTO_CPU_SIM_TARGET_A5` | Compile the A5 CPU instruction paths and use the existing A5 simulator memory profile |
| Existing standalone CPU build with neither flag | Neither defined | Preserve runtime selection and the default A2A3 behavior |

Add `include/pto/cpu/target_config.hpp` to recognize these presence-based
flags. Reject defining both flags, using either without `__CPU_SIM`, or
combining them with NPU target-selection macros. New simpler simulator
builds must always pass exactly one family flag. The no-flag mode preserves
compatibility with other callers and existing CPU tests.

This interface makes no choice between A5 target IDs `3101` and `3510`.
Both use the A5 compute family in the existing ISA architecture mapping,
which is sufficient to select the correct row-sum path. Variant-specific
memory or communication behavior would require a separate, explicit
configuration in a future change; it must not be inferred from `a5sim`.

Do not define `__NPU_ARCH__` or `PTO_NPU_ARCH_A5` to implement this CPU
selection. Those macros also enable NPU headers in `pto_instr_impl.hpp`.
A compile-only probe passes with `__CPU_SIM` alone but fails after adding
`__NPU_ARCH__=3101`, because an NPU header references `atomicAdd`.

#### 2. simpler: derive the family from the existing platform

Pass the existing platform into `Gxx15Toolchain.get_compile_flags` and
append the corresponding family flag. The mapping is direct:

```python
# Proposed flag selection inside get_compile_flags.
target_flags = {
    "a2a3sim": "-DPTO_CPU_SIM_TARGET_A2A3",
    "a5sim": "-DPTO_CPU_SIM_TARGET_A5",
}
flags.append(target_flags[platform])
flags.extend(["-include", "pto/cpu/target_config.hpp"])
```

Reject unsupported simulator platform values with a clear error. Use this
same flag-building path in `KernelCompiler._compile_incore_sim` and
`KernelCompiler.incore_compile_cache_token`, so the cache describes the
actual compilation command.

Example A5 vector-kernel arguments, alongside the existing compiler flags:

```text
-D__CPU_SIM -D__DAV_VEC__ -DPTO_CPU_SIM_TARGET_A5
-include pto/cpu/target_config.hpp
```

For AIC, retain the same family flag and use `__DAV_CUBE__` instead of
`__DAV_VEC__`. Apply the selected family to every translation unit of the
kernel, including externally supplied source kernels. Cache identity and
artifact metadata must distinguish A2A3 from A5. A prebuilt simulator
kernel must declare a matching family or be rebuilt.

The forced include makes a build against an older ISA fail clearly when
the new header is absent, rather than silently ignoring an unknown `-D`
option. Implement the ISA interface first, then advance simpler's managed
ISA pin and enable the flags together. Invalidate older kernel cache
entries. No model changes or generated kernel-entry initialization calls
are needed.

#### 3. pto-isa: specialize instruction semantics and memory together

Resolve the flags to an internal compile-time selector with three modes:
`A2A3`, `A5` and `Legacy`. Pass that selector as a template parameter to
architecture-dependent CPU helpers. For example, row-sum dispatch can use
this structure (illustrative pseudocode):

```cpp
template <CpuSimTarget Target, typename... Args>
void RowSumForTarget(Args&&... args)
{
    if constexpr (Target == CpuSimTarget::A5) {
        RowSumA5(args...);
    } else if constexpr (Target == CpuSimTarget::A2A3) {
        RowSumA2A3(args...);
    } else {
        RowSumUsingRuntimeArch(args...);
    }
}
```

Fixed-family builds must not consult mutable runtime state to select an
algorithm. Preserve the current A5 tree/modular reductions and A2A3
compatibility algorithm, and select their architecture-specific type/layout
checks with the same family. Runtime valid-shape checks remain. Use the
same selector for other family-dependent CPU paths, including `TLOAD`,
and for CPU capability queries.

Use a family-parameterized memory-model implementation, such as
`NPUMemoryModelImpl<Target>`, with separate thread-local state for A2A3,
A5 and legacy mode. On first Tile access, automatically allocate buffers
using the existing simulator profile for that family. Memory allocation
remains a runtime operation; the family and instruction path are selected
at compilation. Keep the existing buffer-size overrides. The A5 profile
is a simulator configuration, not a claim to model every A5 variant's
physical memory limits exactly.

In fixed-family builds, no-argument initialization uses the compiled
family. `Initialize`, `NPU_MEMORY_INIT` and `SetDefaultArch` may accept the
same family, but must reject a different one even with `NDEBUG`.
`GetArch()` must agree with the compiled family. Legacy builds retain
runtime switching and default A2A3 initialization.

Carry the family selector through architecture-dependent instruction
wrappers and helpers as well as the memory model, so their instantiations
have distinct symbols and TLS state. Changing only a macro-dependent
inline default or public type alias is insufficient: kernels compiled for
different families must not resolve to each other's implementation or
state. All translation units in one kernel artifact must use the same
family.

Selecting the A5 family must not automatically define
`PTO_URMA_SUPPORTED`, include NPU communication headers, or enable an
unimplemented CPU engine. Preserve independently supported CPU features;
variant-specific URMA/RDMA support is outside this fix.

#### 4. Delivery and acceptance

Implement and test the ISA interface first, then update the simpler ISA
pin and compiler/cache integration together. Preserve the current minimal
reproduction as the before/after test.

| Check | Acceptance criterion |
| --- | --- |
| A5 minimal reproduction | The default `a5sim` command, without `--initialize-a5` or extra target configuration, returns eight zeros and exits 0 at `atol=rtol=0` |
| A2A3 control | The same input uses the A2A3 path and returns eight ones with the recorded GCC flags; validate against an A2A3-specific reference |
| Compile-time dispatch | Fixed-family row-sum instantiations contain no runtime A2A3/A5 selection; no initialization call is injected into kernel source |
| Memory and API consistency | First access uses the selected family's simulator profile; `GetArch()` agrees; conflicting initialization requests fail even with `NDEBUG` |
| Compiler and cache integration | AIC/AIV commands contain exactly the family flag selected by the platform; otherwise identical A2A3/A5 sources have distinct cache identities and metadata |
| Worker/library isolation | Load A2A3 and A5 kernels in one process in both orders, reuse workers, and verify instruction results and family-specific memory state remain correct |
| Capability boundaries | A5 selection enables its shared compute paths without enabling NPU headers or claiming additional communication support |
| Compatibility | No-flag CPU tests retain runtime switching; conflicting flags, family flags without CPU_SIM, and unsupported ISA versions fail clearly |
| Model replay | Fixed CSA inputs pass all outputs and fixed SWA attention inputs pass `kv_cache`; its independent `x_out` issue remains separate |

### Verification and remaining scope

The standalone control above passes exactly after A5 initialization.
Model controls confirm its effect on the original failures:

| Fixed input | Before initialization | After initialization |
| --- | --- | --- |
| CSA, captured with seed 1 | RMSNorm differs from its reference in 68/57344 BF16 elements; maximum absolute difference 0.0078125 | All 68 differences disappear; all 7 model outputs pass |
| SWA attention, captured with seed 2 | `kv_cache` FAIL, `x_out` FAIL | `kv_cache` PASS, `x_out` still FAIL from the independent synchronization issue |

The pre-initialization CSA values also match the A2A3 sequential-reduction
reference exactly, connecting the model symptom to the selected ISA path.

The recorded results validate runtime initialization as a causal control.
The compile-time interface above has not yet been implemented or validated;
its acceptance checks and model-suite regression remain to be run.
A5 hardware bitwise validation is outside the recorded evidence.

**Comment by Crystal-wzy** [comment]
Fixed in https://github.com/hw-native-sys/simpler/pull/2302.
